src/v/cluster/tx_gateway_frontend.cc

```diff
@@ -1463,46 +1463,48 @@ tx_gateway_frontend::do_commit_tm_tx(
             }
         }
 
-        std::vector<ss::future<prepare_tx_reply>> pfs;
-        for (auto rm : tx.partitions) {
-            pfs.push_back(_rm_partition_frontend.local().prepare_tx(
-              rm.ntp,
-              rm.etag,
-              model::tx_manager_ntp.tp.partition,
-              tx.pid,
-              tx.tx_seq,
-              timeout));
-        }
-
         std::vector<ss::future<prepare_group_tx_reply>> pgfs;
         for (auto group : tx.groups) {
             pgfs.push_back(_rm_group_proxy->prepare_group_tx(
               group.group_id, group.etag, tx.pid, tx.tx_seq, timeout));
         }
 
-        if (
-          !is_transaction_ga()
-          && tx.status == tm_transaction::tx_status::ongoing) {
-            auto preparing_tx = co_await stm->mark_tx_preparing(
-              expected_term, tx.id);
-            if (!preparing_tx.has_value()) {
-                if (preparing_tx.error() == tm_stm::op_status::not_leader) {
-                    outcome->set_value(tx_errc::not_coordinator);
-                    co_return tx_errc::not_coordinator;
+        std::vector<ss::future<prepare_tx_reply>> pfs;
+        auto ok = true;
+        auto rejected = false;
+
+        if (!is_transaction_ga()) {
+            for (auto rm : tx.partitions) {
+                pfs.push_back(_rm_partition_frontend.local().prepare_tx(
+                  rm.ntp,
+                  rm.etag,
+                  model::tx_manager_ntp.tp.partition,
+                  tx.pid,
+                  tx.tx_seq,
+                  timeout));
+            }
+
+            if (tx.status == tm_transaction::tx_status::ongoing) {
+                auto preparing_tx = co_await stm->mark_tx_preparing(
+                  expected_term, tx.id);
+                if (!preparing_tx.has_value()) {
+                    if (preparing_tx.error() == tm_stm::op_status::not_leader) {
+                        outcome->set_value(tx_errc::not_coordinator);
+                        co_return tx_errc::not_coordinator;
```

Comment:
I'm a bit confused about the life cycle of `pfs` and `pgfs` in this function. They are both vectors of futures that are populated here on the stack, but it looks like there are execution paths (e.g. line 1493) in which we `co_return` and let `pfs` and `pgfs` be destroyed. Am I reading this correctly?

If so, then pfs and pgfs can become sets background futures. That it isn't inherently problematic, but it looks like we should set up these background futures to deal with exceptions so that we avoid the potential runtime errors "abandoned exceptional future". Perhaps more importantly, though, this looks like it was inadvertent because the backgrounding is done _implicitly_ as a side effect of returning.

It appears that this issue I'm describing has been around in this code for quite a while, so perhaps @rystsov can comment here?